WIP: VK_VALVE_mutable_descriptor_type

misyltoad opened the pull request:
This extension allows applications to reduce descriptor memory footprint by allowing a
descriptor to be able to mutate to a given list of descriptor types depending on which
descriptor types are written into, or copied into a descriptor set.

The main use case this extension intends to address is descriptor indexing with `VK_DESCRIPTOR_BINDING_VARIABLE_DESCRIPTOR_COUNT_BIT`, as this means applications
can allocate one large descriptor set, rather than having one large descriptor set per
descriptor type, which significantly bloats descriptor memory usage and causes
performance issues.

This extension also adds a mechanism to declare that a descriptor pool, and therefore
the descriptor sets that are allocated from it, reside only in host memory; as such these
descriptors can only be updated/copied, but not bound.

This allows much more efficient emulation of the binding model present in other
graphics APIs for layering or porting purposes.

misyltoad committed:
Initial draft of VK_VALVE_mutable_descriptor_type

HansKristian-Work commented:
Looks pretty good. I think there are three main questions:

- When updating a descriptor that is mutable type, I think we need to be explicitly clear that it **becomes** that type for the matching rules of the API. And when updating the descriptor, it can change into a different type.
- I think to ensure a RADV implementation makes more sense, we should make it so that if HOST_ONLY_BIT pool is not used, then UPDATE_AFTER_BIND_BIT pool must be used along with VK_DESCRIPTOR_BINDING_PARTIALLY_BOUND_BIT_EXT, VK_DESCRIPTOR_BINDING_UPDATE_AFTER_BIND_BIT_EXT, VK_DESCRIPTOR_BINDING_UPDATE_UNUSED_WHILE_PENDING_BIT_EXT descriptor flags. That blocks any mutable push descriptors, which is a nice side effect. BO tracking in general could get very whacky with mutable descriptors, and I'd like to avoid that pitfall.
- I wonder if we should pass VkMutableTypeDescriptorListVALVE to vkCreateDescriptorPool as well?

Joshua-Ashton commented:
> 
> 
> Looks pretty good. I think there are three main questions:
> 
>     * When updating a descriptor that is mutable type, I think we need to be explicitly clear that it **becomes** that type for the matching rules of the API. And when updating the descriptor, it can change into a different type.
Agreed, I will make that more clear.
> 
>     * I think to ensure a RADV implementation makes more sense, we should make it so that if HOST_ONLY_BIT pool is not used, then UPDATE_AFTER_BIND_BIT pool must be used along with VK_DESCRIPTOR_BINDING_UPDATE_AFTER_BIND_BIT_EXT descriptor flag. That blocks any mutable push descriptors, which is a nice side effect. BO tracking in general could get very whacky with mutable descriptors, and I'd like to avoid that pitfall.
I'm a bit confused, do you mean this for without host only + mutable descriptors, or without host only + *copying to* mutable descriptors? Wouldn't they both have the same problems here (potentially)?
> 
>     * I wonder if we should pass VkMutableTypeDescriptorListVALVE to vkCreateDescriptorPool as well?
I'm not sure what this would be for... Descriptor pools don't normally need a list of all descriptor types potentially used in a pool, why would ones that have descriptor sets with mutable types need that...?

HansKristian-Work commented:
> I'm not sure what this would be for... Descriptor pools don't normally need a list of all descriptor types potentially used in a pool, why would ones that have descriptor sets with mutable types need that...?

The only real concern is memory bloat. Say, if you use MUTABLE with just STORAGE_BUFFER and TEXEL_BUFFER, that consumes say 16 bytes, but MUTABLE could be up to 64 bytes.

HansKristian-Work commented:
> I'm a bit confused, do you mean this for without host only + mutable descriptors, or without host only + _copying to_ mutable descriptors? Wouldn't they both have the same problems here (potentially)?

The concern is that we should never have to do BO tracking.
- HOST_ONLY there is no need for tracking to begin with.
- UPDATED_AFTER_BIND pool, there is also no need for tracking.
- A weird case we should avoid is PUSH_DESCRIPTOR with MUTABLE.
- Non-UPDATED_AFTER_BIND with MUTABLE type could be "weird" to allow.

Joshua-Ashton commented:
> 
> 
> > I'm a bit confused, do you mean this for without host only + mutable descriptors, or without host only + _copying to_ mutable descriptors? Wouldn't they both have the same problems here (potentially)?
> 
> The concern is that we should never have to do BO tracking.
> 
>     * HOST_ONLY there is no need for tracking to begin with.
> 
>     * UPDATED_AFTER_BIND pool, there is also no need for tracking.
> 
>     * A weird case we should avoid is PUSH_DESCRIPTOR with MUTABLE.
> 
>     * Non-UPDATED_AFTER_BIND with MUTABLE type could be "weird" to allow.

Got it.

Joshua-Ashton commented:
> 
> 
> > I'm not sure what this would be for... Descriptor pools don't normally need a list of all descriptor types potentially used in a pool, why would ones that have descriptor sets with mutable types need that...?
> 
> The only real concern is memory bloat. Say, if you use MUTABLE with just STORAGE_BUFFER and TEXEL_BUFFER, that consumes say 16 bytes, but MUTABLE could be up to 64 bytes.

Derp, yes. I think we need `VkDescriptorSetLayoutMutableDescriptorTypeCreateInfoVALVE` rather than `VkMutableTypeDescriptorListVALVE` here though (so it can be per-pool size element)

I think we can get away with passing the same `pNext` although, maybe with a less-specific name. :))

oddhack committed:
Change log for October 19, 2020 Vulkan 1.2.158 spec update:

  * Update release number to 158 for this update.

Internal Issues:

  * Clarify that  linear filtering can be used with comparison sampling
    in valid usage statements for dispatched draw commands
    (internal issue 2365).
  * Add valid usage statement for flink:vkGetQueryPoolResults requiring
    pname:stride to be large enough for a single performance query result
    (internal issue 2380).
  * Move input attachment imageLayout to valid usage reference section,
    and refer to the <<attachment-type-imagelayout>> section
    to reduce complexity of valid usage statement
    (internal merge request 4117).
  * Update issues list for `<<VK_QCOM_render_pass_transform>>`
    (internal merge request 4175).
  * Add valid usage statement for
    slink:VkPipelineColorBlendStateCreateInfo::pname:attachmentCount (public
    Vulkan-ValidationLayers issue 2197).

New Extensions:

  * `<<VK_KHR_fragment_shading_rate>>`
  * `<<VK_KHR_shader_terminate_invocation>>`

BNieuwenhuizen commented on chapters/descriptorsets.txt:
```diff
@@ -2190,7 +2307,32 @@ The allocated descriptor sets are returned in pname:pDescriptorSets.
+Descriptors also become undefined: when:
+
+ifdef::VK_VALVE_mutable_descriptor_type[]
+  * Updating a descriptor of ename:VK_DESCRIPTOR_TYPE_MUTABLE_VALVE type,
```
Does it make more sense to make some of these VUs instead of having them make the descriptor undefined. That is a tiny bit more restrictive but would make it likely massively easier to test the added bits in the validation layers?

HansKristian-Work replied:
I guess this can be a VU.

BNieuwenhuizen commented on xml/vk.xml:
```diff
@@ -14163,10 +14178,17 @@ typedef void <name>CAMetalLayer</name>;
             </require>
         </extension>
-        <extension name="VK_EXT_extension_352" number="352" author="EXT" contact="Joshua Ashton @Joshua-Ashton" supported="disabled">
+        <extension name="VK_VALVE_mutable_descriptor_type" number="352" type="device" supported="vulkan" author="VALVE" contact="Joshua Ashton @Joshua-Ashton,Hans-Kristian Arntzen @HansKristian-Work">
```
Should we add a specialuse="d3demulation"? Or do we believe this is general enough to be recommended for use by general vulkan apps?

HansKristian-Work replied:
Should be d3demulation, yes.

oddhack committed:
Change log for November 1, 2020 Vulkan 1.2.159 spec update:

  * Update release number to 159 for this update.

Github Issues:

  * Clarify handle uniqueness with private data in the
    <<fundamentals-objectmodel-overview, Object Model>> section (public
    issue 1349).
  * Make ename:VK_PERFORMANCE_COUNTER_DESCRIPTION_PERFORMANCE_IMPACTING_KHR
    an alias of
    ename:VK_PERFORMANCE_COUNTER_DESCRIPTION_PERFORMANCE_IMPACTING_BIT_KHR
    and ename:VK_PERFORMANCE_COUNTER_DESCRIPTION_CONCURRENTLY_IMPACTED_KHR
    an alias of
    ename:VK_PERFORMANCE_COUNTER_DESCRIPTION_CONCURRENTLY_IMPACTED_BIT_KHR,
    for backwards compatibility while complying with naming conventions
    (public issue 1367).
  * Make ename:VK_SURFACE_COUNTER_VBLANK_EXT an alias of
    ename:VK_SURFACE_COUNTER_VBLANK_BIT_EXT for backwards compatibility
    while complying with naming conventions (public issue 1368).
  * Add a note to the <<memory-model-synchronizes-with, Synchronizes-With>>
    section that fragment shader interlock instructions don't perform
    implicit availability or visibility operations (public issue 1383).

Internal Issues:

  * Dynamically generate SPIR-V <<spirvenv-capabilities-table,
    Capabilities>> and <<spirvenv-extensions-table, Extensions>> tables from
    new tags in `vk.xml`, exposing this information for other projects
    downstream (internal issue 2156).
  * Clarify when a <<renderpass-feedbackloop, feedback loop>> creates a data
    race (internal issue 2296).
  * Remove un-needed `noautovalidity` attributes on pname:pNext structure
    members, including a few cases where they were actually suppressing
    appropriate autogenerated validity statements (internal issue 2335;
    similar purpose to closed public PR 1339).
  * Clarify treatment of most-negative signed normalized fixed-point values
    in the <<fundamentals-fixedfpconv, Conversion from Normalized
    Fixed-Point to Floating-Point>> section (internal issue 2367).
  * Clarify that enabling an extension cannot change existing implementation
    behavior in the introduction to the <<extendingvulkan-extensions,
    Extensions>> chapter (internal issue 2375).
  * Add missing valid usage statement to flink:vkCreatePrivateDataSlotEXT
    (internal issue 2379).
  * Fix a misplaced asciidoctor `endif::` for flink:vkCreateSwapchainKHR
    (internal merge request 4177).
  * Add missing pname:aspectMask valid usage statement to
    slink:VkSubpassDescription2, matching
    slink:VkInputAttachmentAspectReference (internal merge request 4177).
  * Clarify <<interfaces, SPIR-V rules on decorations>> (internal
    spirv/SPIR-V issue 444).
  * Add missing `<<VK_VERSION_1_2>>` to all
    `<<VK_EXT_separate_stencil_usage>>` conditional markup (internal
    vulkansc/vulkansc issue 58).

New Extensions:

  * `<<VK_QCOM_rotated_copy_commands>>` (internal merge request 4132).

oddhack committed:
Change log for November 9, 2020 Vulkan 1.2.160 spec update:

  * Update release number to 160 for this update.

Github Issues:

  * Remove redundant input attachment valid usage statements from
    slink:VkAttachmentReference2 (public issue 1378).

Internal Issues:

  * Restore "`Preamble`" section containing the copyright statement to the
    proper place in the chunked specification index (internal issue 2384).
  * Add missing valid usage statements for
    flink:vkCmdDrawIndirectByteCountEXT (internal issue 2400).
  * Move vertex input example from the <<fxvertex>> chapter to the Vulkan
    Guide (internal merge request 4162).
  * Clarify update-after-bind limits for dynamic buffers in the
    <<limits-maxDescriptorSetUpdateAfterBindUniformBuffersDynamic>> and
    <<limits-maxDescriptorSetUpdateAfterBindStorageBuffersDynamic>> sections
    (internal merge request 4186).
  * Clarify slink:VkFramebufferCreateInfo to allow read-only use of depth
    images as attachments and non-attachments (internal merge request 4191).
  * Remove redundant valid usage statement from slink:VkWriteDescriptorSet
    (internal merge request 4196).
  * Remove redundant valid usage statement from flink:vkFreeDescriptorSets
    (internal merge request 4198).
  * Fix typo in slink:VkDisplaySurfaceCreateInfoKHR valid usage statement
    (internal merge request 4199).
  * Remove redundant pname:firstViewport / pname:firstScissor limit checks
    in valid usage statements for flink:vkCmdSetViewportWScalingNV,
    flink:vkCmdSetScissor, flink:vkCmdSetExclusiveScissorNV,
    flink:vkCmdSetViewportShadingRatePaletteNV, and flink:vkCmdSetViewport
    (internal merge request 4200).

New Extensions:

  * `<<VK_NV_fragment_shading_rate_enums>>`

HansKristian-Work commented on xml/vk.xml:
```diff
@@ -5021,7 +5029,92 @@ typedef void <name>CAMetalLayer</name>;
+        </type>
+        <type category="struct" name="VkMutableTypeDescriptorListVALVE">
+            <member optional="true"><type>uint32_t</type>                                          <name>descriptorTypeCount</name></member>
+            <member len="descriptorTypeCount" optional="false,false"><type>VkDescriptorType</type>* <name>pDescriptorTypes</name></member>
```
This should be const. Ran into const correctness weirdness during impl.
